Add C code generator and runtime2 backend

tuxji opened the pull request:
Add a new runtime2 backend to Daffodil with a C code generator and
libcli/libruntime libraries.  The C code generator implements a small
but still functional subset of DFDL 1.0 for binary data:

- binary boolean, integer, and real numbers
- arrays of simple and complex elements
- choice groups using dispatch/branch keys
- validation of "fixed" attributes
- padding of explicit length complex elements with fill bytes

Also add a new CLI subcommand "daffodil generate" to call the code
generator and modify the CLI subcommand "daffodil test" to support
running TDML tests with daffodil-runtime2.

We are documenting reviewer requested changes in [Runtime2
ToDos](https://daffodil.apache.org/dev/design-notes/runtime2-todos/).
We plan to address these changes while continuing to build out the C
code generator to implement more of the "Runtime 2 Design" table in
<https://cwiki.apache.org/confluence/display/DAFFODIL/WIP%3A+Daffodil+Runtime+2>.
Build requirements and instructions are in the top-level README.md and
BUILD.md files.

DAFFODIL-2202
______________________________________________________________________

Changelog for selected files:

main.yml - Use msys2 shell on windows, install runtime2 build
requirements, use explicit versions for actions and oses.

sonarclound.yml - Install runtime2 build requirements, use explicit
versions for actions and oses.

README.md - Fix or disable markdownlint warnings, wrap paragraphs, add
runtime2 build requirements, sort hyperlink tags alphabetically.

build.sbt - Add runtime2 project, allow environment variables CC and
AR to override cCompiler and ccArchiveCommand's default settings,
ensure C compiler test-builds C files by doing a recursive glob on "c"
and "examples", make tdmlProc and cli depend on runtime2 because they
call it, and add runtime2 to daffodil-test's test classpath.

bin.LICENSE - Add os-lib license.

Main.scala - Add generate subcommand option.

Compiler.scala - Make ProcessorFactory.forLanguage call runtime2 using
Class.forName to avoid core dependency on runtime2.

ElementBase.scala, ElementDeclMixin.scala, ElementRef.scala - Add DSOM
getters for fixedAttr, hasFixedValue, fixedValueAsString & fixedValue,
make IDEA recommended fixes.

ElementBaseGrammarMixin.scala - Make IDEA recommended fixes.

ElementDeclGrammarMixin.scala - Give Runtime2CodeGenerator access to
RootGrammarMixin.documentElement.

Grammar.scala - Give Runtime2CodeGenerator access to
BinaryGram.children.

GrammarTerm.scala - Allow subclasses to override Gram.context and
change its type from SchemaComponent to ElementBase.

Production.scala - Give Runtime2CodeGenerator access to Prod.guard.

ElementCombinator.scala - Give Runtime2CodeGenerator access to
ElementParseAndUnspecifiedLength.{context, eBeforeContext, eValue,
eAfterValue, repTypeElementGram, subComb}, make IDEA recommended
fixes.

PrimitivesBinaryNumber.scala - Give Runtime2CodeGenerator access to
BinaryIntegerKnownLength.{signed, lengthInBits}.

SequenceCombinator.scala - Give Runtime2CodeGenerator access to
OrderedSequence.sequenceChildren, make IDEA recommended fixes.

tdml.xsd - Add daffodil-runtime2 to tdml:{defaultImplementations,
implementations}.

dafext.xsd - Add tdmlImplementation to tunables.

DFDLParserUnparser.scala - Add ProcessorFactory.forLanguage, trait
CodeGenerator, split half of DataProcessor into DataProcessorBase.

DataProcessor.scala - Minor cleanups.

TDMLRunner.scala - Create TDMLDFDLProcessorFactory or
Runtime2TDMLDFDLProcessorFactory object depending on
tdmlImplementation tunable and call cleanUp on parse/unparse results.

TDMLDFDLProcessor.scala - Add TDMLResult.cleanUp abstract function.

DaffodilTDMLDFDLProcessor.scala - Implement
DaffodilTDMLParseResult.cleanUp and DaffodilTDMLUnparseResult.cleanUp.

Runtime2TDMLDFDLProcessor.scala - Run TDML tests on runtime2 backend.
Use cleanUp to remove all temporary files.

choice-unparse2.tdml - Change model name from "c" to "c1" to avoid
conflict with daffodil-runtime2's "c" resources directory.

Dependencies.scala - Add os-lib dependency for writing C files.

Rat.scala - Ignore runtime2 tests' binary data files.

plugins.sbt - Add sbt-cc plugin so sbt can test-build C files.

sonar-project.properties - Make scan ignore C files since SonarCloud
CLI Docker action can't analyze them anyway.  Also prevent some scan
warnings.

mbeckerle commented:
This builds and tests perfectly from sbt at a command shell using Java 16.
I'm not having success with IntelliJ IDEA though.  Just trying to build (or rebuild) daffodil project gives "Module 'daffodil-udf' tests:java.lang.ExceptionInInitializerError". Which typically reflects some linkage problem. 

Any suggestions? Does IntelliJ IDEA need to use an earlier Java version perhaps?

tuxji commented:
I verified my IDEA has no problem rebuilding the daffodil project (Build -> Rebuild Project or Build -> Build Project, even after typing "clean" in IDEA's interior sbt shell pane) when my default JDK is Java 11 (sdk use java 11.0.10.hs-adpt).

SDKMAN! has 9 versions of Java 16 available; I picked the AdoptOpenJDK Hotspot one, installed it, and made it my default JDK.

```shell
interran@GH3WPL13E:~$ sdk list java | rg 16
 AdoptOpenJDK  |     | 16.0.0.j9    | adpt    |            | 16.0.0.j9-adpt
               |     | 16.0.0.hs    | adpt    |            | 16.0.0.hs-adpt
 Amazon        |     | 16.0.0.36.1  | amzn    |            | 16.0.0.36.1-amzn
 Azul Zulu     |     | 16.0.0       | zulu    |            | 16.0.0-zulu
               |     | 16.0.0.fx    | zulu    |            | 16.0.0.fx-zulu
 BellSoft      |     | 16.0.0.fx    | librca  |            | 16.0.0.fx-librca
               |     | 16.0.0       | librca  |            | 16.0.0-librca
               |     | 16           | open    |            | 16-open
 SAP           |     | 16           | sapmchn |            | 16-sapmchn
interran@GH3WPL13E:~$ sdk install java 16.0.0.hs-adpt

Downloading: java 16.0.0.hs-adpt

In progress...

#################################################################################################################################################################################### 100.0%

Repackaging Java 16.0.0.hs-adpt...

Done repackaging...

Installing: java 16.0.0.hs-adpt
Done installing!

Do you want java 16.0.0.hs-adpt to be set as default? (Y/n): y

Setting java 16.0.0.hs-adpt as default.
interran@GH3WPL13E:~$
```

I relaunched my IDEA using Java 16 and had no problem rebuilding the daffodil project again (Build -> Rebuild Project).  Here's what my sbt shell pane showed:

```
/home/interran/.sdkman/candidates/java/current/bin/java -server -Xmx1536M -Dsbt.supershell=false -Didea.managed=true -Dfile.encoding=UTF-8 -Didea.installation.dir=/home/interran/.apps/idea-IC -jar /home/interran/.local/share/JetBrains/IdeaIC2021.1/Scala/launcher/sbt-launch.jar early(addPluginSbtFile=\"\"\"/tmp/idea.sbt\"\"\") "; set ideaPort in Global := 46073 ; idea-shell"
[info] welcome to sbt 1.5.0 (AdoptOpenJDK Java 16)
[info] loading global plugins from /home/interran/.sbt/1.0/plugins
[info] compiling 1 Scala source to /home/interran/.sbt/1.0/plugins/target/scala-2.12/sbt-1.0/classes ...
[info] Non-compiled module 'compiler-bridge_2.12' for Scala 2.12.13. Compiling...
[info]   Compilation completed in 7.292s.
[info] done compiling
[info] loading settings for project daffodil-runtime2-build-build-build from metals.sbt ...
[info] loading project definition from /home/interran/apache/daffodil-runtime2/project/project/project
[info] loading settings for project daffodil-runtime2-build-build from metals.sbt ...
[info] loading project definition from /home/interran/apache/daffodil-runtime2/project/project
[success] Generated .bloop/daffodil-runtime2-build-build.json
[success] Total time: 0 s, completed Apr 19, 2021, 11:49:00 AM
[info] loading settings for project daffodil-runtime2-build from metals.sbt,plugins.sbt,idea.sbt ...
[info] loading project definition from /home/interran/apache/daffodil-runtime2/project
[success] Generated .bloop/daffodil-runtime2-build.json
[info] compiling 2 Scala sources to /home/interran/apache/daffodil-runtime2/project/target/scala-2.12/sbt-1.0/classes ...
[info] done compiling
[success] Total time: 1 s, completed Apr 19, 2021, 11:49:03 AM
[info] loading settings for project daffodil from build.sbt ...
[info] loading settings for project daffodil-japi from build.sbt ...
[info] loading settings for project daffodil-cli from build.sbt ...
[info] resolving key references (26028 settings) ...
[info] set current project to daffodil (in build file:/home/interran/apache/daffodil-runtime2/)
[info] Defining Global / ideaPort
[info] The new value will be used by Compile / compile, Test / compile and 37 others.
[info] 	Run `last` for details.
[info] Reapplying settings...
[info] set current project to daffodil (in build file:/home/interran/apache/daffodil-runtime2/)
[IJ]all {file:/home/interran/apache/daffodil-runtime2/}daffodil-io/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-io/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-japi/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-japi/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-cli/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-cli/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-core/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-core/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil/products {file:/home/interran/apache/daffodil-runtime2/}daffodil/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-sapi/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-sapi/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-schematron/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-schematron/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-runtime1-unparser/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-runtime1-unparser/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-runtime2/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-runtime2/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-propgen/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-propgen/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-runtime1/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-runtime1/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-lib/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-lib/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-macro-lib/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-macro-lib/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-udf/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-udf/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-test-stdLayout/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-test-stdLayout/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-tutorials/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-tutorials/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-test/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-test/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-test-ibm1/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-test-ibm1/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-tdml-lib/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-tdml-lib/test:products {file:/home/interran/apache/daffodil-runtime2/}daffodil-tdml-processor/products {file:/home/interran/apache/daffodil-runtime2/}daffodil-tdml-processor/test:products
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/resource_managed/main/org/apache/daffodil/xsd/dafext.xsd
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/resource_managed/main/org/apache/daffodil/xsd/dfdlx.xsd
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/resource_managed/main/org/apache/daffodil/xsd/DFDL_part3_model.xsd
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/resource_managed/main/org/apache/daffodil/xsd/DFDL_part2_attributes.xsd
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/resource_managed/main/org/apache/daffodil/xsd/DFDL_part1_simpletypes.xsd
[info] compiling 6 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-macro-lib/target/scala-2.12/classes ...
[info] compiling 3 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-propgen/target/scala-2.12/classes ...
[info] compiling 5 Java sources to /home/interran/apache/daffodil-runtime2/daffodil-udf/target/scala-2.12/classes ...
[info] Non-compiled module 'compiler-bridge_2.12' for Scala 2.12.11. Compiling...
[info] done compiling
[info] compiling 9 Scala sources and 23 Java sources to /home/interran/apache/daffodil-runtime2/daffodil-udf/target/scala-2.12/test-classes ...
[info]   Compilation completed in 7.655s.
[info] done compiling
[info] compiling 1 Scala source to /home/interran/apache/daffodil-runtime2/daffodil-propgen/target/scala-2.12/test-classes ...
[info] done compiling
[info] done compiling
[info] done compiling
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/src_managed/main/org/apache/daffodil/schema/annotation/props/gen/GeneratedCode.scala
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/src_managed/main/org/apache/daffodil/api/DaffodilTunablesGen.scala
[info] generated /home/interran/apache/daffodil-runtime2/daffodil-lib/src_managed/main/org/apache/daffodil/api/WarnIdGen.scala
[info] compiling 73 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-lib/target/scala-2.12/classes ...
[info] done compiling
[info] compiling 41 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-io/target/scala-2.12/classes ...
[info] compiling 41 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-lib/target/scala-2.12/test-classes ...
[info] compiling 5 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-schematron/target/scala-2.12/classes ...
[info] done compiling
[info] done compiling
[info] compiling 31 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-io/target/scala-2.12/test-classes ...
[info] compiling 5 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-tdml-lib/target/scala-2.12/classes ...
[info] compiling 162 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-runtime1/target/scala-2.12/classes ...
[info] done compiling
[info] done compiling
[info] done compiling
[info] compiling 3 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-tdml-lib/target/scala-2.12/test-classes ...
[info] done compiling
[info] done compiling
[info] compiling 12 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-runtime1/target/scala-2.12/test-classes ...
[info] compiling 37 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-runtime1-unparser/target/scala-2.12/classes ...
[info] done compiling
[info] done compiling
[info] compiling 114 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-core/target/scala-2.12/classes ...
[info] done compiling
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli/xml_reader.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/xml_reader.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli/daffodil_argp.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/daffodil_argp.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] compiling 47 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-core/target/scala-2.12/test-classes ...
[info] compiling 6 Scala sources and 6 Java sources to /home/interran/apache/daffodil-runtime2/daffodil-japi/target/scala-2.12/classes ...
[info] compiling 10 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-sapi/target/scala-2.12/classes ...
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli/stack.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/stack.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli/xml_writer.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/xml_writer.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli/daffodil_main.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/daffodil_main.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime/parsers.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/parsers.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime/errors.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/errors.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime/infoset.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/infoset.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime/unparsers.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/unparsers.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples/NestedUnion.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/examples/NestedUnion.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [cc, -c, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples/ex_nums.c, -o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/examples/ex_nums.c.o, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libcli, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/c/libruntime, -I/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/main/resources/examples, -Wall, -Wextra]
[info] [ar, cr, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccTargetMap/_global/streams/compile/sbtcc.Library/libruntime2.a, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/xml_reader.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/daffodil_argp.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/stack.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/xml_writer.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libcli/daffodil_main.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/parsers.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/errors.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/infoset.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/c/libruntime/unparsers.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/examples/NestedUnion.c.o, /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/streams/compile/ccSourceObjectMap/_global/streams/compile/libruntime2.a/src/main/resources/examples/ex_nums.c.o]
[info] compiling 11 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/scala-2.12/classes ...
[info] done compiling
[info] compiling 9 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-sapi/target/scala-2.12/test-classes ...
[info] compiling 6 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-schematron/target/scala-2.12/test-classes ...
[info] done compiling
[info] done compiling
[info] compiling 12 Java sources to /home/interran/apache/daffodil-runtime2/daffodil-japi/target/scala-2.12/test-classes ...
[info] done compiling
[info] done compiling
[info] compiling 5 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-tdml-processor/target/scala-2.12/classes ...
[info] done compiling
[info] done compiling
[info] compiling 2 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-cli/target/scala-2.12/classes ...
[info] compiling 11 Scala sources and 1 Java source to /home/interran/apache/daffodil-runtime2/daffodil-tdml-processor/target/scala-2.12/test-classes ...
[info] compiling 3 Scala sources to /home/interran/apache/daffodil-runtime2/test-stdLayout/target/scala-2.12/test-classes ...
[info] compiling 8 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-test-ibm1/target/scala-2.12/test-classes ...
[info] compiling 1 Scala source and 1 Java source to /home/interran/apache/daffodil-runtime2/tutorials/target/scala-2.12/test-classes ...
[info] done compiling
[info] done compiling
[info] done compiling
[info] done compiling
[info] done compiling
[info] compiling 1 Scala source to /home/interran/apache/daffodil-runtime2/daffodil-cli/target/scala-2.12/test-classes ...
[info] done compiling
[info] done compiling
[info] compiling 1 Scala source to /home/interran/apache/daffodil-runtime2/daffodil-runtime2/target/scala-2.12/test-classes ...
[info] done compiling
[info] compiling 129 Scala sources to /home/interran/apache/daffodil-runtime2/daffodil-test/target/scala-2.12/test-classes ...
[info] done compiling
[success] Total time: 58 s, completed Apr 19, 2021, 11:50:25 AM
[IJ]eval System.getProperty("java.version")
[info] ans: String = 16
[IJ]eval System.getProperty("java.home")
[info] ans: String = /home/interran/.sdkman/candidates/java/16.0.0.hs-adpt
[IJ]

>
```

Which operating system, JDK vendor, and IDEA version are you using?  I'm using Ubuntu 20.04.2 LTS running in WSL 2 on Windows 10.

mbeckerle commented:
My IntelliJ IDEA is 2021.1
These difficulties seem to have evaporated. I am not sure why.

mbeckerle reviewed (approved):
+1

stevedlawrence commented on project/Dependencies.scala:
```diff
@@ -22,6 +22,7 @@ object Dependencies {
   lazy val common = core ++ infoset ++ test
 
   lazy val core = Seq(
+    "com.lihaoyi" %% "os-lib" % "0.7.1", // for generating C source files
```
Latest version of 0.7.3. Might as well update now instead of waiting for this to merge then getting a PR from scala steward.

stevedlawrence reviewed (commented):
Looks good! Handful of minor comments, only real concern is about some category X licenses.

mbeckerle commented on daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dafext.xsd:
```diff
@@ -504,6 +504,14 @@
+          <xs:annotation>
+            <xs:documentation>
+              TDMLDFDLProcessorFactory implementation to use when running TDML tests.
+              Allowed values are "daffodil" (default), "daffodil-runtime2", and "ibm".
```
I suggest extensibility here is a good "feature request" but wouldn't do anything other than what is here for now.

stevedlawrence replied:
I noticed later that we actually do hardcode these values in, so enumerating them here very reasonable, if not necessary. Ignore my suggestion.

tuxji reviewed (commented):
I've responded to all the comments and made all the suggested changes.  This pull request will be ready for a second review after the checks pass.

mbeckerle commented on daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:
```diff
@@ -502,11 +502,47 @@ class CLIConf(arguments: Array[String]) extends scallop.ScallopConf(arguments)
+    footer("""|
+              |Run 'daffodil generate <language> --help' for subcommand specific options""".stripMargin)
+
+    object c extends scallop.Subcommand("c") {
```
"c-runtime2" would be good. There's definitely no reason to be trying to describe characteristics of the C code (thread safe or not etc.) in these identifiers, but the kind of thing being generated is C code, with runtime 2 in mind. 
I'm ignoring that "runtime2" is itself a bit challenged as a name, but I've long believed that sometimes the best distinction between things is just "add a number" because trying to capture description in a short identifier is hard.

stevedlawrence replied:
"c-runtime2" makes me think there's a c-runtime1. Maybe just plain "c" is reasonable, considering [DAFFODIL-2406](https://issues.apache.org/jira/browse/DAFFODIL-2406) suggests renaming things to "backend-c-generator". The "generator" part is assumed in the "daffodil generator" subcommand, and backend is also implied. So maybe just plain "c" is the best option?

mbeckerle replied:
Agree that "c" is fine for now. No change required here.

mbeckerle commented on daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala:
```diff
@@ -477,7 +476,13 @@ abstract class TestCase(testCaseXML: NodeSeq, val parent: DFDLTestSuite)
+      // Right now daffodil and ibm use the same ProcessFactory name
+      case "daffodil" | "ibm" => "org.apache.daffodil.tdml.processor.TDMLDFDLProcessorFactory"
+      case "daffodil-runtime2" => "org.apache.daffodil.tdml.processor.Runtime2TDMLDFDLProcessorFactory"
+      case other => Assert.invariantFailed("'%s' not valid for tdmlImplementation".format(other))
```
Add this concept to the TODO asciidoc page? (If not already there.)

tuxji replied:
I had already mentioned my first idea in the runtime2 ToDos ascii page but not this second idea.  I'll submit a pull request to update the ToDos.

stevedlawrence replied:
One thing to point out about IBM cross tester and adding it permanently to the class path is that IBM DFDL and Daffodil have conflicting dependencies. I believe IBM depends on ICU 51 and Daffodil depends on a much newer version. I don't remember if these differences just cause certain tests to fail or if there are actual binary incompatibilities that cause other issues. But I think that's the reaoson why instructions require manually enabling/disabling the ibm cross tester.

I think there's been mention that Java 9 modules might help with this, but I haven't looked into it in much detail.

stevedlawrence commented on .github/workflows/main.yml:
```diff
@@ -25,30 +25,64 @@ jobs:
+        if: runner.os == 'Windows'
+        uses: msys2/setup-msys2@v2
+        with:
+          install: gcc libargp-devel make pkgconf
```
As to libargb being widely distributed, I dont' think that matters to ASF policy. The [license page](https://www.apache.org/legal/resolved.html) says the following about dependencies to components with restricted licenses:

> Apache projects can rely on components under prohibited licenses if the component is only needed for optional features. When doing so, a project shall provide the user with instructions on how to obtain and install the non-included work. Optional means that the component is not required for standard use of the product or for the product to achieve a desirable level of quality. The question to ask yourself in this situation is:
>
> "Will the majority of users want to use my product without adding the optional components?"

So we just need to answer this question.

From my perspective, if we generate code, I would think most users would expect that code to compile to the daffodil/daffodil.exe binary and then use that binary, even if libargp weren't installed. If a user doesn't have that installed, I suspect right now compilation will fail?

Perhaps if we could split things up so we don't build the CLI by default, I think that makes it more clear that the CLI is not required and not expected to be used, and that additional actions are required. I would say by default, tests shouldn't fail if the dependency isn't installed unless the user wants to explicitly enable this feature.

I would also argue that if we truly don't expect users to use the CLI, then we don't need all the capabilities that libargp provides. Short options that something like getopt provides seems reasonable from a functionality perspective.

mbeckerle replied:
Actually, we could write around libargp. Probably we should.

I'm more concerned about depending on gcc specifically. 

Are we depending on gcc specifically as opposed to any other C compiler?

stevedlawrence replied:
I think we only depend on a C99 capable compiler. So users can decide whatever compiler they want that meets their needs.    That said, it might make sense to use clang in all our examples and github actions. It should work just the same and ASF would probably be more happy with that.

tuxji replied:
We depend only on a C99 compliant compiler.  I've used clang as well as gcc on Ubuntu 20.04 and I can verify it works just the same.  I can update BUILD.md and the github actions to use clang although I'd have to find the correct steps and test they work on multiple platforms (e.g., install the clang-10 package and set CC=clang-10 and AR=llvm-ar-10 on Ubuntu 20.04).  Do you want those more complicated steps in the BUILD.md and the github actions now or as a TODO?  Likewise for replacing argp with getopt?  I'd prefer a TODO for the latter at least.

mbeckerle replied:
Well, I would say these need to be fixed before any release, but you could do them in a separate JIRA ticket and separate PR.

stevedlawrence replied:
Agreed, that's my only hesitation with this PR. As long as it's fixed before the 3.1 release I'm a +1

stevedlawrence commented on .github/workflows/main.yml:
```diff
@@ -25,30 +25,64 @@ jobs:
+          make install
+          # Workaround for sbt hanging problem
+          echo "COURSIER_CACHE=$temp" >> $GITHUB_ENV
+          echo "COURSIER_CONFIG_DIR=$temp" >> $GITHUB_ENV
```
Gotcha. Is there a bug or anything we can reference to track this issue with dev.dirs? And is this specific to mxml installation? Wondering if maybe this cousier stuff be done in a different step?

tuxji replied:
No, the dev.dirs hanging problem has nothing to do with mxml installation.  I put the >> $GITHUB_END commands there only because I needed a step that runs only on Windows and lets me change the environment before the sbt compile step.  

I need to file a bug report with a reproducible test case in dev.dirs.  Note that dev.dirs exhibits the problem but they may or may not be responsible for it.  Their code which tries to run a Windows PowerShell script using a Java subprocess call hangs when run from MSYS2 on Windows although it works fine when run from CMD on Windows.

tuxji commented:
@stevedlawrence, I've updated the runtime2 todos, created two new JIRA issues, rebased to fix a conflict with the master branch, and squashed the runtime2 commits.  Please approve if everything looks good.

stevedlawrence commented:
+1 from me! Glad to get this merged in!!
